design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+
+```yaml
+name: Ensure Event counts are within bounds
+checks:
+  - id: great_expectations_validate_events
+    type: webhook
+    properties:
+      url: "http://<host:port>/checks/great_expectations"
+      query_params:
+        condition: expect_table_row_count_to_be_between(min_value=2000, max_value=5000)
+      headers:
+        secret_header: "{{ ENV.MY_SECRET }}"
+```
+
+#### Check event syntax
+
+```json
+{
+  "repository_id": "example-repo",
+  "branch_id": "branch_a",
+  "source_ref": "0307e79ab5afef2b0726a8c13896f716033b1acd8454e22210bd43a2f72a94a9",
+  "check_id": "great_expectations_validate_events",
+  "storage_namespace": "s3://path/to/storage/",
+  "callback_token": "abc123",
+  "output_url": "presigned URL or another HTTP POST-able URL"
+}
+```
+
+#### Check triggering
+
+Checks are triggered when the `check(repository_id, ref_id)` endpoint is called in the lakeFS API.
+When running check() with a branch or tag, it will be dereferenced: checks are valid for a given commit ID.
+
+
+#### Check execution
+
+Like hooks, checks could have multiple type implementations (Airflow, Webhooks, etc). Initially, only Airflow is supported.
+
+`execute()` will trigger a new [DAG run using the Airflow REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dag_runs), passing as conf the settings as shown above (including the callback token)
+
+#### Check status
+
+```mermaid
+stateDiagram-v2
+    direction LR
+    [*] --> STARTING: execute()
+    STARTING --> FAILED: error calling Airflow
+    STARTING --> EXECUTING
+    EXECUTING --> SUCCESS: callback(success)
+    EXECUTING --> FAILED: callback(error)
+    EXECUTING --> LOST: timeout
+    LOST --> STARTING: retry
+    FAILED --> STARTING: retry
+    SUCCESS --> [*]
+```
+
+#### Retrying checks
+
+From the `lost` and `failed` statuses, users can trigger a retry operationthat will cause the execute function to run again for the given check.
+This is used when transient errors occur (or a timeout).
+
+#### Getting check status
+
+##### `/api/v1/repositories/{repoID}/refs/{refID}/checks`
+
+```json
+{
+  "checks": [
+    {"id": "great_expectations_validate_events", "status": "SUCCESS", "execution_id": "1234"},
+    {"id": "other_check", "status": "EXECUTING", "execution_id": "45678"},
+    {"id": "yet_another_check", "status": "LOST", "execution_id": "8900"},
+  ]
+}
+```
+
+#### Callback API
+
+Each execution will send a `callback_token` to the executing system (Airflow, for now).
+
+`POST /api/v1/repositories/{repoID}/refs/{refID}/checks/{checkID}?token={callback_token}`
+
+```json
+{
+  "status": "SUCCESS",
+  "metadata": {},
+}
+```
+
+Where metadata is a simple `Map<string, string>` mapping to be sent by the executing system.
+
+When lakeFS receives a callback, it checks that the provided token is the last one issued for that (check, commit). 
+Any previous token issued will be ignored.
+
+#### Merging into a protected branch
```

Comment:
Successful merges to the protected branch may introduce / remove checks to the list
These check might have not been run by potential merge branches or might have failed during their checks execution.
How do we plan to handle these kind of situations? What is the wanted behavior in these cases?

Reply:
validation of required checks is done on the destination branch pre-merge. If a merge removes or changes the required checks, the new state will have to conform to whatever was merged. Does that make sense?